cpp/src/quantiles/quantiles.cu

```diff
@@ -61,48 +60,216 @@ quantiles(table_view const& input,
                           mr);
 }
 
-}
+template<typename DataIterator, typename Result>
+struct quantiles_data_functor
+{
+    DataIterator data;
+    size_type size;
+    interpolation interp;
+
+    Result __device__ operator()(double q) {
+        return detail::select_quantile_data<Result>(data, size, q, interp);
+    }
+};
+
+template<typename ValidityIterator>
+struct quantiles_mask_functor
+{
+    ValidityIterator validity;
+    size_type size;
+    interpolation interp;
+
+    bool __device__  operator() (double q) {
+        return detail::select_quantile_validity(validity, size, q, interp);
+    }
+};
+
+
+template<typename SortMapIterator>
+struct quantiles_functor
+{
+
+    SortMapIterator ordered_indices;
+    std::vector<double> const& q;
+    interpolation interp;
+    bool retain_types;
+    rmm::mr::device_memory_resource* mr;
+    cudaStream_t stream;
+
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_arithmetic<T>::value, std::unique_ptr<column>>
+    operator()(Args&&... args)
+    {
+        CUDF_FAIL("Only arithmetic types are supported in quantiles.");
```

Comment:
I think this message needs to be clearer. Only arithmetic types are supported for arithmetic interpolation schemes.